fedimint-client/src/lib.rs

```diff
@@ -131,50 +228,93 @@ impl Debug for ClientInner {
     }
 }
 
+/// Global state given to a specific client module and state. It is aware inside
+/// which module instance and operation it is used and to avoid module being
+/// aware of their instance id etc.
+#[derive(Clone, Debug)]
+struct ModuleGlobalClientContext {
+    client: Arc<ClientInner>,
+    module_instance: ModuleInstanceId,
+    operation: OperationId,
+}
+
 #[apply(async_trait_maybe_send!)]
-impl IGlobalClientContext for ClientInner {
+impl IGlobalClientContext for ModuleGlobalClientContext {
     fn api(&self) -> &(dyn IFederationApi + 'static) {
-        self.api.as_ref()
+        self.client.api.as_ref()
     }
 
-    async fn finalize_and_submit_transaction(
+    async fn claim_input_dyn(
         &self,
         dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
-        operation_id: OperationId,
-        tx_builder: TransactionBuilder,
+        input: InstancelessDynClientInput,
+    ) -> TransactionId {
+        let instance_input = ClientInput {
+            input: DynInput::from_parts(self.module_instance, input.input),
+            keys: input.keys,
+            state_machines: states_add_instance(self.module_instance, input.state_machines),
+        };
+
+        self.client
+            .finalize_and_submit_transaction(
```

Comment:
Is the general guidance that new clients shouldn't call `finalize_and_submit_transaction` directly and instead should call one of these wrappers? (`claim_input`, `fund_output`)

Reply:
When constructing transactions using module client-generated inputs and outcputs there is only support for `finalize_and_submit_transaction` (that's implemented on `Client`) right now, inside state transitions only the `claim_input` and `fund_output` functions can be used, the `finalize_and_submit_transaction` isn't exposed anymore.

This will become clearer once I add the last commit to #2270